homeassistant/components/snapcast/media_player.py

```diff
@@ -58,36 +57,23 @@
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the Snapcast platform."""
 
+    _LOGGER.debug("Reached async_setup_platform")
+
     host = config.get(CONF_HOST)
     port = config.get(CONF_PORT, CONTROL_PORT)
 
-    async def async_service_handle(service_event, service, data):
-        """Handle dispatched services."""
-        entity_ids = data.get(ATTR_ENTITY_ID)
-        devices = [
-            device for device in hass.data[DATA_KEY] if device.entity_id in entity_ids
-        ]
-        for device in devices:
-            if service == SERVICE_SNAPSHOT:
-                device.snapshot()
-            elif service == SERVICE_RESTORE:
-                await device.async_restore()
-            elif service == SERVICE_JOIN:
-                if isinstance(device, SnapcastClientDevice):
-                    master = [
-                        e
-                        for e in hass.data[DATA_KEY]
-                        if e.entity_id == data[ATTR_MASTER]
-                    ]
-                    if isinstance(master[0], SnapcastClientDevice):
-                        await device.async_join(master[0])
-            elif service == SERVICE_UNJOIN:
-                if isinstance(device, SnapcastClientDevice):
-                    await device.async_unjoin()
-
-        service_event.set()
-
-    async_dispatcher_connect(hass, DOMAIN, async_service_handle)
+    platform = entity_platform.current_platform.get()
+    platform.async_register_entity_service(SERVICE_SNAPSHOT, {}, "snapshot")
+    platform.async_register_entity_service(SERVICE_RESTORE, {}, "async_restore")
+    platform.async_register_entity_service(
+        SERVICE_JOIN, {vol.Required(ATTR_MASTER): cv.entity_id}, "async_join"
```

Comment:
There will be an attribute error if the user sends an entity_id or a master id for an entity that is not a client but a group. The old code guarded for that. Do we need to do that still or at least raise a more specific error?

We could do that by using the alternative entity service handler api. See the bottom paragraph here:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

Eg:
```py
async def handle_async_join(entity, service_call):
	if not isinstance(entity, SnapcastClientDevice):
        raise ValueError("Incorrect entity_id. Entity is not a client.")
    await entity.async_join(service_call.data[ATTR_MASTER])
```

Then also do a similar check in the entity `async_join` method for the master entity.

Reply:
Still working through these...sorry for the extra request